thread_pthread_mn.c

```diff
@@ -674,12 +713,30 @@ timer_thread_register_waiting(rb_thread_t *th, int fd, enum thread_sched_waiting
 }
 
 static void
-timer_thread_unregister_waiting(rb_thread_t *th, int fd)
+timer_thread_unregister_waiting(rb_thread_t *th, int fd, enum thread_sched_waiting_flag flags)
 {
     RUBY_DEBUG_LOG("th:%u fd:%d", rb_th_serial(th), fd);
-
+#if HAVE_SYS_EVENT_H
+    if (flags) {
+        struct kevent ke[2];
+        int num_events = 0;
+
+        if (flags & thread_sched_waiting_io_read) {
+            EV_SET(&ke[num_events], fd, EVFILT_READ, EV_DELETE, 0, 0, NULL);
+            num_events++;
+        }
+        if (flags & thread_sched_waiting_io_write) {
+            EV_SET(&ke[num_events], fd, EVFILT_WRITE, EV_DELETE, 0, 0, NULL);
+            num_events++;
+        }
```

Comment:
Without verifying the flags involved, `kevent` can result in errors removing filters that were not added.